### PR TITLE
fix: [NPM-WIN] lock the policyMap & updatePodCache

### PR DESCRIPTION
**Pull request by huntergregory**
When applying lots of pod events and policy events at once, there's a chance of a fatal concurrent write to a map from the both the Pod Controller and NetPol controller. _**This is strictly a Windows issue**_ because in Linux, only the NetPol controller's code path will touch the PolicyManager, but in Windows, we also add/remove any necessary policies to a pod's endpoint when the pod's IPSets are updated.
Now we lock the policyMap to prevent this from happening.

This PR also locks the updatePodCache (_**only used by Windows**_) because sometimes we would update a pod twice. We haven't seen crashes for this before, but there could be a chance of concurrent map access here too, and either way, we shouldn't make a repeated OS call.
ApplyDataplane() can be called by the Pod controller and NetPol controller code paths, so in the old code, a pod could be updated while the other thread is waiting for HNS to respond.

### CODE REVIEW

File: npm/pkg/dataplane/dataplane_windows.go [hunk 1]
```diff
@@ -349,8 +357,8 @@ func (dp *DataPlane) setNetworkIDByName(networkName string) error {
 }
 
 func (dp *DataPlane) getAllEndpointIDs() []string {
-	endpointIDs := make([]string, 0, len(dp.endpointCache))
-	for _, endpoint := range dp.endpointCache {
+	endpointIDs := make([]string, 0, len(dp.endpointCache.cache))
```
Comment (vakalapa) [inline npm/pkg/dataplane/dataplane_windows.go:360]: epCache lock unlock here.

File: npm/pkg/dataplane/dataplane_windows.go [hunk 2]
```diff
@@ -118,7 +118,7 @@ func (dp *DataPlane) updatePod(pod *updateNPMPod) error {
 	}
 
 	// Check if pod is already present in cache
-	endpoint, ok := dp.endpointCache[pod.PodIP]
+	endpoint, ok := dp.endpointCache.cache[pod.PodIP]
```
Comment (vakalapa) [inline npm/pkg/dataplane/dataplane_windows.go:121]: Need to lock and unlock here

File: npm/pkg/dataplane/dataplane.go [hunk 3]
```diff
@@ -23,6 +24,24 @@ type Config struct {
 	*policies.PolicyManagerCfg
 }
 
+type updatePodCache struct {
```
Comment (vakalapa) [inline npm/pkg/dataplane/dataplane.go:27]: Move both updatepodcache and endpointcache to types.go and have Get, Set functions which take care of lock unlock inside.

File: npm/pkg/dataplane/policies/policymanager.go [hunk 4]
```diff
@@ -36,14 +36,45 @@ type PolicyManagerCfg struct {
 }
 
 type PolicyMap struct {
+	mutex sync.RWMutex
 	cache map[string]*NPMNetworkPolicy
 }
 
+func (m *PolicyMap) RLock() {
+	if util.IsWindowsDP() {
```
Comment (vakalapa) [inline npm/pkg/dataplane/policies/policymanager.go:44]: Dont need this check,

Comment (huntergregory) [reply npm/pkg/dataplane/policies/policymanager.go]: removed

File: npm/pkg/dataplane/dataplane.go (hunk 3 above)
Comment (huntergregory) [reply npm/pkg/dataplane/dataplane.go:27]: discussed offline that this is irrelevant now

File: npm/pkg/dataplane/dataplane_windows.go (hunk 2 above)
Comment (huntergregory) [reply npm/pkg/dataplane/dataplane_windows.go:120]: fixed locking

File: npm/pkg/dataplane/dataplane_windows.go (hunk 1 above)
Comment (huntergregory) [reply npm/pkg/dataplane/dataplane_windows.go:372]: this is used only at bootup, so no need to use locks

**Review by vakalapa — APPROVED** [review]

**Pull request merged**